File: utils/validation.go
package utils

import (
	"net/http"
	"strings"

	apierror "github.com/akrck02/valhalla-core-sdk/error"

	"github.com/akrck02/valhalla-core-sdk/lang"
	apimodels "github.com/akrck02/valhalla-core-sdk/models/api"
)

const MINIMUM_CHARACTERS_FOR_PASSWORD = 16
const MINIMUM_CHARACTERS_FOR_EMAIL = 5

func ValidateEmail(email string) *apimodels.Error {

	if len(email) < MINIMUM_CHARACTERS_FOR_EMAIL {
		return &apimodels.Error{
			Status:  http.StatusBadRequest,
			Error:   apierror.ShortEmail,
			Message: "Email must have at least " + lang.Int2String(MINIMUM_CHARACTERS_FOR_EMAIL) + " characters",
		}
	}

	if !strings.Contains(email, "@") {
		return &apimodels.Error{
			Status:  http.StatusBadRequest,
			Error:   apierror.NoAtEmail,
			Message: "Email must have one @",
		}
	}

	if !strings.Contains(email, ".") {
		return &apimodels.Error{
			Status:  http.StatusBadRequest,
			Error:   apierror.NoDotEmail,
			Message: "Email must have at least one .",
		}
	}

	return nil
}

func ValidatePassword(password string) *apimodels.Error {

	if len(password) < MINIMUM_CHARACTERS_FOR_PASSWORD {
		return &apimodels.Error{
			Status:  http.StatusBadRequest,
			Error:   apierror.ShortPassword,
			Message: "Password must have at least " + lang.Int2String(MINIMUM_CHARACTERS_FOR_PASSWORD) + " characters",
		}
	}

	if !ContainsSpecialCharacters(password) {
		return &apimodels.Error{
			Status:  http.StatusBadRequest,
			Error:   apierror.NoSpecialCharactersPassword,
			Message: "Password must have at least one special character",
		}
	}

	if IsLowerCase(password) {
		return &apimodels.Error{
			Status:  http.StatusBadRequest,
			Error:   apierror.NoMayusMinusPassword,
			Message: "Password must have at least one uppercase character",
		}
	}

	if IsUpperCase(password) {
		return &apimodels.Error{
			Status:  http.StatusBadRequest,
			Error:   apierror.NoMayusMinusPassword,
			Message: "Password must have at least one lowercase character",
		}
	}

	if !ContainsNumbers(password) {
		return &apimodels.Error{
			Status:  http.StatusBadRequest,
			Error:   apierror.NoAlphanumericPassword,
			Message: "Password must have at least one number",
		}
	}

	return nil
}

func ValidateName(name string) *apimodels.Error {

	if len(name) < 2 {
		return &apimodels.Error{
			Status:  http.StatusBadRequest,
			Error:   apierror.ShortName,
			Message: "Name must have at least 2 characters",
		}
	}

	if len(name) > 50 {
		return &apimodels.Error{
			Status:  http.StatusBadRequest,
			Error:   apierror.LongName,
			Message: "Name must have at most 50 characters",
		}
	}

	return nil
}

func ValidateDescription(description string) *apimodels.Error {

	if len(description) < 2 {
		return &apimodels.Error{
			Status:  http.StatusBadRequest,
			Error:   apierror.ShortDescription,
			Message: "Description must have at least 2 characters",
		}
	}

	if len(description) > 500 {
		return &apimodels.Error{
			Status:  http.StatusBadRequest,
			Error:   apierror.LongDescription,
			Message: "Description must have at most 500 characters",
		}
	}

	return nil
}

File: src/main/kotlin/org/akrck02/valhalla/core/sdk/validation/Validation.kt
package org.akrck02.valhalla.core.sdk.validation

import org.akrck02.valhalla.core.sdk.error.ErrorCode
import org.akrck02.valhalla.core.sdk.model.exception.ServiceException
import org.akrck02.valhalla.core.sdk.model.http.HttpStatusCode

const val MINIMUM_CHARACTERS_FOR_PASSWORD = 16
const val MINIMUM_CHARACTERS_FOR_EMAIL = 5

/**
 * Validate string as an email to match the requirements
 * @throws ServiceException if the email is invalid
 */
@Suppress("unused")
fun String.validateEmail() {

    takeIf { it.isNotBlank() } ?: throw ServiceException(
        status = HttpStatusCode.BadRequest,
        code = ErrorCode.InvalidRequest,
        message = "Email is empty."
    )

    takeIf { it.length >= MINIMUM_CHARACTERS_FOR_EMAIL } ?: throw ServiceException(
        status = HttpStatusCode.BadRequest,
        code = ErrorCode.InvalidEmail,
        message = "Email must have at least $MINIMUM_CHARACTERS_FOR_EMAIL characters."
    )

    takeIf { it.contains("@") } ?: throw ServiceException(
        status = HttpStatusCode.BadRequest,
        code = ErrorCode.InvalidEmail,
        message = "Email must have one @."
    )

    takeIf { it.contains(".") } ?: throw ServiceException(
        status = HttpStatusCode.BadRequest,
        code = ErrorCode.InvalidEmail,
        message = "Email must have at least one dot."
    )

}

/**
 * Validate string as a password to match the requirements
 * @throws ServiceException if the password is invalid
 */
@Suppress("unused")
fun String.validatePassword() {

    takeIf { it.isNotBlank() } ?: throw ServiceException(
        status = HttpStatusCode.BadRequest,
        code = ErrorCode.InvalidPassword,
        message = "Password is empty."
    )

    takeIf { it.length >= MINIMUM_CHARACTERS_FOR_PASSWORD } ?: throw ServiceException(
        status = HttpStatusCode.BadRequest,
        code = ErrorCode.InvalidPassword,
        message = "Password must have at least $MINIMUM_CHARACTERS_FOR_PASSWORD characters."
    )

    takeIf { it.isLowercase().not().and(it.isUppercase().not()) } ?: throw ServiceException(
        status = HttpStatusCode.BadRequest,
        code = ErrorCode.InvalidPassword,
        message = "Password must have at least one lowercase and uppercase character."
    )

    takeIf { it.containsSpecialCharacters() } ?: throw ServiceException(
        status = HttpStatusCode.BadRequest,
        code = ErrorCode.InvalidPassword,
        message = "Password must have at least one special character."
    )

    takeIf { it.containsNumbers() } ?: throw ServiceException(
        status = HttpStatusCode.BadRequest,
        code = ErrorCode.InvalidPassword,
        message = "Password must have at least one number."
    )
}

/**
 * Get if a String is lowercase
 * @return true if it is, false otherwise
 */
fun String.isLowercase(): Boolean = this.lowercase() == this

/**
 * Get if a String is uppercase
 * @return true if it is, false otherwise
 */
fun String.isUppercase(): Boolean = this.uppercase() == this

/**
 * Get if a String contains numbers
 * @return true if it is, false otherwise
 */
fun String.containsNumbers(): Boolean = this.contains(Regex("[0-9]+"))

/**
 * Get if a String contains special characters
 * @return true if it is, false otherwise
 */
fun String.containsSpecialCharacters(): Boolean = this.contains(Regex("[¡!@#$%^&¿?*]+"))
